pkgs/native_assets_builder/doc/protocol/v1.md

```diff
@@ -0,0 +1,231 @@
+Build Hook Specification
+========================
+
+Version 1: As of 2024-01-24 implemented behind `--enable-experiment=native-assets`
+
+### Concepts
+
+#### Asset
+
+An asset is a file together with metadata that tells the Dart and Flutter SDK how to bundle such asset in an application bundle.
+
+Currently, all such assets are dynamic libraries.
+
+An asset must have an `assetId`. Dart code that uses an asset, references the asset using the `assetId`.
+
+An asset must have a target os and target architecture. When accessing the asset at runtime, the asset for the current os and architecture is accessed.
+
+An asset must have a link mode. Currently, the only supported link mode is dynamic. (Static linking of native code will be added later.)
+
+An asset must specify a path type for dynamic linking. 
+
+* `absolute` paths are absolute paths in the output dir (see below).
+* `system` paths are expected to resolve on the target machine PATH. In this case the asset is not a file but only metadata.
+* `process` "paths" have no path, symbols are resolved in the current process. In this case the asset is not a file but only metadata.
+* `executable` "paths" have no path, symbols are resolved in the current executable. In this case the asset is not a file but only metadata.
+
+#### AssetId
+An asset must have an `assetId`. Dart code that uses an asset, references the asset using the `assetId`.
+
+A package must prefix all assetIds it defines with: `package:<package>/`, this ensures assets don't conflict between packages.
+
+Conventionally, an asset referenced from `lib/src/foo.dart` in `package:foo` has `assetId = 'package:foo/src/foo.dart'`.
+
+#### Target os and architecture
+
+A target specifies the operating system and architecture of the targeted platform.
+
+### `build.dart`
+To bundle native assets with a package, that package must define a `build.dart`
+script which accepts a `--config` option as follows:
+
+```console
+$ dart build.dart --config <build_config.yaml>
+```
+
+The `--config` option specifies location of the `build_config.yaml` file.
+
+The `build.dart` file MUST:
+ * Read the `build_config.yaml` file,
+ * Build assets using configuration from `build_config.yaml`.
+   * If `dry_run: true` in `build_config.yaml`, then this may be skipped.
+ * Write assets into `out_dir` from `build_config.yaml`.
+   * If `dry_run: true` in `build_config.yaml`, then this may be skipped.
+   * Filename are unrelated to `assetId`.
+   * Arbitrary file names are fine, `build_output.yaml` will map `assetId` to files. 
+   * MUST avoid file name `build_output.yaml`.
+ * Write `build_output.yaml` into `out_dir` (from `build_config.yaml`).
+   * This maps `assetId` to assets previous written in `out_dir`.
+   * There may be multiple assets for a given `assetId` depending on
```

Comment:
You might want to add something about the uniqueness of an `assetId` - this was unclear in an earlier review comment as well.